File: Algoritmos/doubly_linked_list.go
package main

import "fmt"

type node struct {
	elem int
	next *node
	prev *node
}

type list struct {
	size uint
	head *node
	tail *node
}

func new_node(elem int, next *node, prev *node) *node {
	return &node {elem, next, prev}
}

func create_list() *list {
	l := list {
		size: 0,
		head: new_node(0, nil, nil),
		tail: new_node(0, nil, nil),
	}

	l.head.next = l.tail
	l.tail.prev = l.head

	return &l
}

func (l *list) begin() *node {
	return l.head.next
}

func (l *list) end() *node {
	return l.tail
}

func (l *list) r_begin() *node {
	return l.tail.prev
}

func (l *list) r_end() *node {
	return l.head
}

func (l *list) push_back(elem int) {
	n := new_node(elem, l.tail, l.tail.prev)

	n.prev.next = n
	l.tail.prev = n

	l.size++
}

func (l *list) push_front(elem int) {
	n := new_node(elem, l.head.next, l.head)

	n.next.prev = n
	l.head.next = n

	l.size++
}

func (l *list) pop_back(error) int {
	elem := l.tail.prev.elem

	l.tail.prev = l.tail.prev.prev
	l.tail.prev.next = l.tail

	l.size--
	return elem
}

func (l *list) pop_front() int {
	elem := l.head.next.elem

	l.head.next = l.head.next.next
	l.head.next.prev = l.head

	l.size--
	return elem
}

func (l *list) pop_node(n *node) {
	n.prev.next = n.next
	n.next.prev = n.prev
	l.size--
}

func (l *list) empty() bool {
	return l.head.next == l.tail 
}

func (l *list) remove(elem int) {

	for n := l.begin(); n != l.end(); {

		if n.elem == elem {
			n = n.next
			l.pop_node(n.prev)
		} else {
			n = n.next
		}
	}
}

func (l *list) clear() {
	for l.size > 0 {
		l.pop_front()
	}
}

func (l *list) print_list() {

	for n := l.begin(); n != l.end(); n = n.next {
		fmt.Printf("%v ", n.elem)
	}
	fmt.Println()
}

func main() {
	var l list = *create_list()

	l.push_back(10)
	l.push_back(11)
	l.push_back(12)
	l.push_front(9)
	l.push_back(10)
	l.push_back(21)

	l.remove(10)
	l.print_list()

	l.clear()
	l.print_list()
}

File: Algoritmos/Listas/doublyLinkedList.py
class Node:
    def __init__(self, data, next: "Node" = None, prev: "Node" = None, origin: "List" = None) -> None:
        self.data = data
        self.next = next
        self.prev = prev
        self.origin = origin

class List:
    def __init__(self) -> None:
        self.size = int(0)

        self.head = Node(0, None, None, self)
        self.tail = Node(0, None, None, self)

        self.head.next = self.tail
        self.tail.prev = self.head
    
    def is_empty(self) -> bool:
        return self.size == 0
    
    def is_in_list(self, node: Node) -> bool:
        return node.origin == self
    
    def push_back(self, data) -> None:
        n = Node(data, self.tail, self.tail.prev, self)
        
        self.tail.prev.next = n
        self.tail.prev = n
        self.size += 1
    
    def push_front(self, data) -> None:
        n = Node(data, self.head.next, self.head, self)
        
        self.head.next.prev = n
        self.head.next = n
        self.size += 1

    def pop_back(self) -> None:
        if self.size == 0:
            raise IndexError("Cannot pop from an empty List")
        
        data = self.tail.prev.data

        self.tail.prev.prev.next = self.tail
        self.tail.prev = self.tail.prev.prev
        self.size -= 1

        return data
    
    def pop_front(self) -> None:
        if self.size == 0:
            raise IndexError("Cannot pop from an empty List")
        
        data = self.head.next.data

        self.head.next.next.prev = self.head
        self.head.next = self.head.next.next
        self.size -= 1
    
        return data

    def pop_node(self, node: Node) -> None:
        if node == None:
            raise ValueError("A node must be provided")

        if not self.is_in_list(node):
            raise RuntimeError("The node does not belong to the List")
        
        if node == self.head or node == self.tail:
            raise IndexError("The provided node is out of range")
        
        node.next.prev, node.prev.next = node.prev, node.next
        self.size -= 1

    def clear(self) -> None:
        while self.size != 0:
            self.pop_back()

    def __iter__(self):
        curr = self.head.next

        while curr != self.tail:
            yield curr.data
            curr = curr.next
    
    def __reversed__(self):
        curr = self.tail.prev

        while curr != self.head:
            yield curr.data
            curr = curr.prev
